Do not run Python during 'configure_ros2'

bluka4 opened the pull request:
- Add possibility to provide user-defined `setup.bzl`
- Generate 'setup.bzl' as part of `repos_lock.update`

bluka4 commented on repos/config/detail/BUILD.bazel:
```diff
@@ -11,15 +11,22 @@
+    args = [
+        "$(execpath @ros2//:ros2.repos)",
+        "$(execpath {})".format(REPOS_LOCK_FILE),
+        "$(execpath {})".format(REPOS_SETUP_FILE),
```
It is dirty hacked together. I will clean up if the approach, in general, is going the right way.

A user should provide the `setup.bzl` file from the client repo. During `repos_lock.update`, it will be populated with repository-loading content.

An example how it will look like on the client side:

```python
configure_ros2(
    distro = "humble",
    repo_index_overlays = ["@my_prj//thirdparty/rules_ros_vendor:overlay.repos"],  # optional overlay
    setup_file = "@my_prj//thirdparty/rules_ros_vendor:setup_ros_repositories.bzl",
)
```

kilian-funk commented:
@bluka4 The approach is the right. Please be aware that the setup.bzl file needs to be specific to the distro. So we will need a setup_<distro>.bzl file for each <distro>.
Also, to get this rolling, you will need to have the created files added to the PR. Currently this is broken.

bluka4 commented:
> @bluka4 The approach is the right. Please be aware that the setup.bzl file needs to be specific to the distro. So we will need a setup_.bzl file for each . Also, to get this rolling, you will need to have the created files added to the PR. Currently this is broken.

@kilian-funk, I have updated PR according to your suggestions. Please take a look. The pipeline is green now.

Instead of `*.lock` files, we have `*.lock.bzl` files for each ROS 2 distribution. They are generated by the `repos_lock.update` script based on `ros2.repos` files pulled from the official ROS 2 release bundle.

Alongside `configure_ros2,` I've added the `configure_repos` macro, which allows the client to provide a custom input `.repos` file instead of the official ROS 2 one.

kilian-funk commented on repos/config/detail/generate_ros2_config.py:
```diff
@@ -17,7 +16,9 @@
 
 
-HEADER = """
+HEADER = """#
```
```suggestion
HEADER = """# DO NOT EDIT THIS FILE MANUALLY!
```

kilian-funk reviewed (changes requested):
LGTM I left some minor change requests in the code.

bluka4 commented on WORKSPACE:
```diff
@@ -18,7 +18,7 @@ load("//repos/config:defs.bzl", "configure_ros2")
 configure_ros2(distro = "humble")
 
-load("@ros2_config//:setup.bzl", "setup")
+load("//repos/config:setup_humble.lock.bzl", "setup")
```
Nice! Thanks, @kilian-funk! I was looking into how to remove this duplication but failed.

bluka4 commented on repos/config/detail/ros2_config.bzl:
```diff
@@ -15,35 +15,25 @@
+    ctx.file("repos_overlay_files.bzl", content = "REPOS_OVERLAY_FILES = {}".format(["{}".format(l) for l in ctx.attr.repos_index_overlays]))
+    ctx.file("repos_setup_file.bzl", content = "REPOS_SETUP_FILE = '{}'".format(ctx.attr.setup_file))
     ctx.file("WORKSPACE", content = "workspace(name = {}".format(ctx.name), executable = False)
     ctx.file("BUILD", executable = False)
```
Thanks! This makes it much better! I've applied your suggestion. Please take a look.
